tests: lwm2m: Add more tests to RD client, engine and registry

SeppoTakalo opened the pull request:
### tests: lwm2m: Add bootstrap tests into RD client tests

Add bootstrap and few corner cases into RD client tests
to improve code coverage.

### tests: lwm2m: Add engine tests on security and socket opts

Add tests regarding socket options and various security
options.

### tests: lwm2m: Add more tests to LwM2M registry

Add module tests for all data types and most of deprecated
functions as well.

SeppoTakalo committed:
tests: lwm2m: Add bootstrap tests into RD client tests

Add bootstrap and few corner cases into RD client tests
to improve code coverage.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: lwm2m: Add engine tests on security and socket opts

Add tests regarding socket options and various security
options.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: lwm2m: Add more tests to LwM2M registry

Add module tests for all data types and most of deprecated
functions as well.

Refactor event wainting to expect_lwm2m_rd_client_event() which
does not care about the index, just the order.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lwm2m: Ensure string termination when reading empty data

When resource data length is set to zero, we still need to
ensure that string reading will return an empty string.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/3e15b1d0-e627-43bc-aca3-b9c77d46ec5a)

Slight improvement to code coverage.

SeppoTakalo commented:
Previous values:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/1e4db82a-2d3a-4a06-9b53-eccfa73c0b4b)

There is still way to go...

ssharks commented on subsys/net/lib/lwm2m/lwm2m_registry.c:
```diff
@@ -1156,8 +1156,8 @@ static int lwm2m_engine_get(const struct lwm2m_obj_path *path, void *buf, uint16
-			((char *)buf)[buflen - 1] = '\0';
+			strncpy(buf, data_ptr, data_len - 1);
+			((char *)buf)[data_len - 1] = '\0';
 			break;
```
Is this an intended change, where is ensured that data_len <= buflen?

rlubos replied:
There's a generic check `lwm2m_check_buf_sizes()` above.

SeppoTakalo replied:
@ssharks Yes, this is intentional.

buflen refers to a buffer size where we are going to write. `data_len` is the amount of data we have in registry.
If binary write to string buffer has been made, I still need to ensure that my `strncpy()` does not write more than what we have stored.

So if we write 10 bytes of garbage without terminating character, then ask for read of 20 bytes, previously we got 20 bytes of garbage. So the limiting size value should be the `data_len` variable. Not `buflen`.

Buffer sizes are already verified in `lwm2m_check_buf_sizes()`.

ssharks reviewed (commented):
I'm very new to LWM2M, so on contents it is hard to judge, just one code thing I found

rlubos commented on tests/net/lib/lwm2m/lwm2m_rd_client/src/main.c:
```diff
@@ -496,8 +556,47 @@ ZTEST(lwm2m_rd_client, test_suspend_resume_registration)
+	zassert_true(lwm2m_rd_client_resume() == 0, NULL);
+	zassert_true(check_lwm2m_rd_client_event(LWM2M_RD_CLIENT_EVENT_DISCONNECT, 3), NULL);
+}
+
```
nit: double empty line.

rlubos commented on tests/net/lib/lwm2m/lwm2m_registry/src/test_obj.c:
```diff
@@ -0,0 +1,110 @@
+time_t t;
+double d;
+struct lwm2m_objlnk objl;
+
```
nit: double empty lines.

rlubos commented on tests/net/lib/lwm2m/lwm2m_registry/src/lwm2m_registry.c:
```diff
@@ -352,6 +385,158 @@ ZTEST(lwm2m_registry, test_strings)
 }
 
+#if CONFIG_TEST_DEPRECATED
+/* Don't test deprecated functions on Twister builds */
```
Not sure I follow - what's the point of adding this test then if we don't want to run it with the CI?

SeppoTakalo replied:
It allows us to test those functions manually. I don't think they will change much, so I don't think it is relevant to keep running those in CI.

But I just wrote the tests once to verify and check for code coverage. I have no strong opinion whether I should drop those, or keep for manual testing purposes.

rlubos replied:
If it's already implemented and you plan to use it for the time being I think it could stay - just please remember to remove the test cases when we decide to remove the deprecated APIs.

rlubos reviewed (commented):
Looks good, but please rebase - there's a conflict

SeppoTakalo commented:
Rebased and fixed the whitespace errors.

SeppoTakalo commented:
* Rebased on top of main
* Fixed broken tests
* Refactored `check_lwm2m_rd_client_event()` to `expect_lwm2m_rd_client_event(()` which should be more stable as it does not assume certain index of the event, just check that order is correct and certain event have happened since last call.
